api/odilo.py

```diff
@@ -0,0 +1,495 @@
+
+    def get_hold(self, patron, pin, record_id):
+        patron_holds = self.get_patron_holds(patron, pin)
+        for hold in patron_holds:
+            if hold['recordId'] == record_id and hold['status'] in ('informed', 'waiting'):
+                return hold
+
+        raise NotFoundOnRemote("Could not find active hold for patron %s, record %s" % (patron, record_id))
+
+    def fulfill(self, patron, pin, licensepool, internal_format):
+        record_id = licensepool.identifier.identifier
+        content_link, content, content_type = self.get_fulfillment_link(patron, pin, record_id, internal_format)
+
+        if not content_link and not content:
+            self.log.info("Odilo record_id %s was not available as %s" % (record_id, internal_format))
+        else:
+            return FulfillmentInfo(
+                licensepool.collection,
+                DataSource.ODILO,
+                Identifier.ODILO_ID,
+                record_id,
+                content_link=content_link,
+                content=content,
+                content_type=content_type,
+                content_expires=None
+            )
+
+    def get_fulfillment_link(self, patron, pin, record_id, format_type):
+        """Get the link corresponding to an existing checkout.
+        """
+        # Retrieve checkout with its download_ulr. It is necessary to generate a download token in our API
+        checkout = self.get_checkout(patron, pin, record_id)
+        loan_format = checkout['format']
+        if format_type and loan_format and (
+                        format_type == loan_format or
+                        (format_type == 'EBOOK_STREAMING' and loan_format in ('EPUB', 'PDF'))
+        ):
+            if 'downloadUrl' in checkout and checkout['downloadUrl']:
+                content_link = checkout['downloadUrl']
+                content = None
+                content_type = OdiloRepresentationExtractor.format_data_for_odilo_format[format_type]
+
+                # Get also .acsm file
+                if 'ACSM' == format_type:
+                    response = self.patron_request(patron, pin, content_link)
+                    if response.status_code == 200:
+                        content = response.content
+                    elif response.status_code == 404 and response.content:
+                        error_id, error_description = self.raise_exception_on_error(response.json())
+                        raise CannotFulfill(error_id + ': ' + error_description)
+
+                return content_link, content, content_type
+
+        raise CannotFulfill("Cannot obtain a download link for patron[%r], record_id[%s], format_type[%s].", patron,
+                            record_id, format_type)
+
+    def get_patron_checkouts(self, patron, pin):
+        data = self.patron_request(patron, pin, self.PATRON_CHECKOUTS_ENDPOINT.format(patronId=patron)).json()
+        self.raise_exception_on_error(data)
```

Comment:
This assumes raise_exception_on_error will raise an exception, which isn't the case for your implementation. Same comment for line 286.